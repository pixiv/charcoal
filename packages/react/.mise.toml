[tools]
pnpm = "10.17.1"
