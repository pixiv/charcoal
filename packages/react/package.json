{
  "name": "@charcoal-ui/react",
  "version": "4.0.0-beta.3",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs.js",
      "import": "./dist/index.esm.js",
      "default": "./dist/index.esm.js"
    },
    "./dist/index.css": {
      "import": "./dist/index.css",
      "require": "./dist/index.css"
    }
  },
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "run-p --print-label 'build:*'",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "typecheck": "tsc --project tsconfig.build.json --pretty --noEmit",
    "clean": "rimraf dist .tsbuildinfo"
  },
  "devDependencies": {
    "@charcoal-ui/esbuild-plugin-styled-components": "^4.0.0-beta.3",
    "@react-types/switch": "^3.1.2",
    "@storybook/addon-actions": "^7.4.1",
    "@storybook/addon-knobs": "^7.0.2",
    "@storybook/addons": "^7.4.1",
    "@storybook/api": "^7.4.1",
    "@storybook/components": "^7.4.1",
    "@storybook/core-events": "^7.4.1",
    "@storybook/react": "^7.4.1",
    "@storybook/react-webpack5": "^7.4.1",
    "@storybook/theming": "^7.4.1",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.0",
    "@types/jest-axe": "^3.5.3",
    "@types/react": "^17.0.38",
    "@types/react-dom": "^17.0.11",
    "@types/react-router-dom": "^5.3.3",
    "@types/styled-components": "^5.1.21",
    "@types/warning": "^3.0.0",
    "jest-axe": "^5.0.1",
    "jest-styled-components": "^7.0.8",
    "npm-run-all": "^4.1.5",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^6.2.1",
    "rimraf": "^3.0.2",
    "storybook": "^7.4.1",
    "styled-components": "^5.3.3",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@charcoal-ui/icons": "^4.0.0-beta.3",
    "@charcoal-ui/styled": "^4.0.0-beta.3",
    "@charcoal-ui/theme": "^4.0.0-beta.3",
    "@charcoal-ui/utils": "^4.0.0-beta.3",
    "@react-aria/button": "^3.9.1",
    "@react-aria/checkbox": "^3.13.0",
    "@react-aria/dialog": "^3.5.10",
    "@react-aria/focus": "^3.16.0",
    "@react-aria/link": "^3.6.5",
    "@react-aria/overlays": "^3.20.0",
    "@react-aria/radio": "^3.10.0",
    "@react-aria/ssr": "^3.9.1",
    "@react-aria/switch": "^3.6.0",
    "@react-aria/textfield": "^3.14.1",
    "@react-aria/utils": "^3.23.0",
    "@react-aria/visually-hidden": "^3.8.8",
    "polished": "^4.1.4",
    "react-spring": "^9.0.0",
    "react-stately": "^3.26.0",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": ">=17.0.0",
    "styled-components": ">=5.1.1"
  },
  "files": [
    "src",
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/react"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
