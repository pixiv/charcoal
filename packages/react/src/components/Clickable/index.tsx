import React from 'react'
import styled, { css } from 'styled-components'
import {
  LinkProps,
  useComponentAbstraction,
} from '../../core/ComponentAbstraction'
import { disabledSelector } from '@charcoal/utils'

interface BaseProps {
  /**
   * クリックの無効化
   */
  disabled?: boolean
}

interface LinkBaseProps {
  /**
   * リンクのURL。指定するとbuttonタグではなくaタグとして描画される
   */
  to: string
}

export type ClickableProps =
  | (BaseProps & Omit<React.ComponentPropsWithoutRef<'button'>, 'disabled'>)
  | (BaseProps & LinkBaseProps & Omit<LinkProps, 'to'>)
export type ClickableElement = HTMLButtonElement & HTMLAnchorElement

const Clickable = React.forwardRef<ClickableElement, ClickableProps>(
  function Clickable(props, ref) {
    const { Link } = useComponentAbstraction()
    if ('to' in props) {
      const { onClick, disabled = false, ...rest } = props
      return (
        <A<typeof Link>
          {...rest}
          as={disabled ? undefined : Link}
          onClick={disabled ? undefined : onClick}
          aria-disabled={disabled}
          ref={ref}
        />
      )
    } else {
      return <Button {...props} ref={ref} />
    }
  }
)
export default Clickable

const clickableCss = css`
  /* Clickable style */
  cursor: pointer;

  ${disabledSelector} {
    cursor: default;
  }
`

const Button = styled.button`
  /* Reset button appearance */
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;

  &:focus {
    outline: none;
  }

  /* Change the font styles in all browsers. */
  font: inherit;

  /* Remove the margin in Firefox and Safari. */
  margin: 0;

  /* Show the overflow in Edge. */
  overflow: visible;

  /* Remove the inheritance of text transform in Firefox. */
  text-transform: none;

  /* Remove the inner border and padding in Firefox. */
  &::-moz-focus-inner {
    border-style: none;
    padding: 0;
  }

  ${clickableCss}
`

const A = styled.span`
  /* Reset a-tag appearance */
  color: inherit;

  &:focus {
    outline: none;
  }

  .text {
    top: calc(1em + 2em);
  }

  ${clickableCss}
`
