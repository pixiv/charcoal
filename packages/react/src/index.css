:root {
  /* charcoal custom css variables */
  --charcoal-color-custom-focus-effect: 0 0 0 4px rgba(0, 150, 250, 0.32);
  --charcoal-color-custom-error-effect: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
