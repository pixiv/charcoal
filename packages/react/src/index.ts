export {
  default as ComponentAbstraction,
  useComponentAbstraction,
  type LinkProps,
} from './core/ComponentAbstraction'
export { SSRProvider } from './core/SSRProvider'
export { default as Button, type ButtonProps } from './components/Button'
export {
  default as Clickable,
  type ClickableProps,
  type ClickableElement,
} from './components/Clickable'
export {
  default as IconButton,
  type IconButtonProps,
} from './components/IconButton'
export {
  default as Radio,
  type RadioProps,
  RadioGroup,
  type RadioGroupProps,
} from './components/Radio'
export {
  default as Select,
  type SelectProps,
  SelectGroup,
  type SelectGroupProps,
} from './components/Select'
export { default as Switch, type SwitchProps } from './components/Switch'
export {
  default as TextField,
  type TextFieldProps,
  type SingleLineTextFieldProps,
  type MultiLineTextFieldProps,
} from './components/TextField'
export { default as Icon, type IconProps } from './components/Icon'
export { default as Modal, type ModalProps } from './components/Modal'
export {
  default as LoadingSpinner,
  LoadingSpinnerIcon,
} from './components/LoadingSpinner'
export {
  default as SegmentedControl,
  type SegmentedControlProps,
} from './components/SegmentedControl'
export { default as Checkbox } from './components/Checkbox'
export { default as TagItem, type TagItemProps } from "./components/TagItem"
