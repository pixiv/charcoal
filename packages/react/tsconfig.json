{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "types": ["vitest/globals", "@testing-library/jest-dom"],
    "isolatedDeclarations": true,
  },
  "include": ["./src", "./docs", "./vitest.setup.ts"]
}
