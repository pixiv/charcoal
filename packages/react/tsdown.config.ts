import { defineConfig } from 'tsdown'
// import { styledComponentsPlugin } from '@charcoal-ui/esbuild-plugin-styled-components'
// import postcss from 'postcss'
// import fs from 'node:fs/promises'
// import path from 'node:path'

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['esm', 'cjs'],
  // outExtension({ format }) {
  //   return {
  //     js: `.${format}.js`,
  //   }
  // },
  // banner: {
  //   js: `"use client";`,
  // },
  target: 'esnext',
  sourcemap: true,
  dts:true,
  // minify:true,
  
  // esbuildPlugins: [
  //   styledComponentsPlugin,
  //   {
  //     name: 'at-layer-charcoal',
  //     setup(build) {
  //       // NOTE: this will be called twice for esm and cjs.
  //       build.onEnd(async (result) => {
  //         const originalCssOutput = result.outputFiles?.find((file) =>
  //           file.path.endsWith('css')
  //         )
  //         if (!originalCssOutput) {
  //           throw new Error('[at-layer-charcoal]: expect original css output')
  //         }

  //         const layeredCssFilePath = originalCssOutput.path.replace(
  //           'index.css',
  //           'layered.css'
  //         )
  //         const layeredCssOutput = await postcss({
  //           postcssPlugin: 'at-layer-charcoal',
  //           Once(_, { AtRule, result }) {
  //             const atLayerNode = new AtRule({
  //               name: 'layer',
  //               params: 'charcoal',
  //               nodes: result.root.nodes,
  //             })
  //             result.root.nodes = [atLayerNode]
  //           },
  //         }).process(originalCssOutput.text, {
  //           from: originalCssOutput.path,
  //           to: layeredCssFilePath,
  //           map: {
  //             inline: false,
  //           },
  //         })

  //         // https://github.com/evanw/esbuild/issues/2999
  //         // this will be called before dist is created
  //         await fs.mkdir(path.dirname(layeredCssFilePath), { recursive: true })
  //         await Promise.all([
  //           fs.writeFile(layeredCssFilePath, layeredCssOutput.content),
  //           fs.writeFile(
  //             `${layeredCssFilePath}.map`,
  //             layeredCssOutput.map.toString()
  //           ),
  //         ])

  //         // eslint-disable-next-line no-console
  //         console.log(
  //           `${build.initialOptions.format?.toUpperCase()} dist/layered.css`
  //         )
  //       })
  //     },
  //   },
  // ],
})
