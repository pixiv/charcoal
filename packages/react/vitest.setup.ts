import '@testing-library/jest-dom/vitest'
import { vi } from 'vitest'

vi.mock('react', async (importOriginal) => ({
  useId: vi.fn(() => 'test-id'),
  ...(await importOriginal()),
}))
