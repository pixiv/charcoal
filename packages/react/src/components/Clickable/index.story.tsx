import { action } from '@storybook/addon-actions'
import Clickable from '.'
import { Meta, StoryObj } from '@storybook/react'

export default {
  title: 'react/Clickable',
  component: Clickable,
} as Meta<typeof Clickable>

export const Button: StoryObj = {
  render: () => (
    <Clickable
      style={{
        color: 'var(--charcoal-color-text-secondary-default)',
      }}
      onClick={action('click')}
    >
      button
    </Clickable>
  ),
}

export const Link: StoryObj = {
  render: () => (
    <Clickable
      style={{
        color: 'var(--charcoal-color-text-secondary-default)',
      }}
      component="a"
      href="#"
      onClick={action('click')}
    >
      link
    </Clickable>
  ),
}
