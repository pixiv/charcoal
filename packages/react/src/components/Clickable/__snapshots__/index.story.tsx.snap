// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clickable Button smoke-test 1`] = `
.c0 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

<div data-dark="false">
  <button class="c0">
    button
  </button>
</div>
`;

exports[`Clickable Link smoke-test 1`] = `
.c0 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

<div data-dark="false">
  <a href="#"
     class="c0"
  >
    link
  </a>
</div>
`;
